prow/gangway/gangway.proto

```diff
@@ -0,0 +1,124 @@
+syntax = "proto3";
+
+import "google/api/annotations.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "k8s.io/test-infra/prow/gangway";
+
+service Prow {
+  // FIXME: In the future we can just return a unique token (only), in the same
+  // way that GCB returns immediately with the globally-unique BuildId. That is,
+  // in the future the response will be a union of either the full JobExecution
+  // message or a single JobExecutionToken (string). See
+  // https://docs.google.com/document/d/1v77jp1Nb5C2C2-PdV02SGViO9CyZ9SvNxCPOHyIUQeo/edit#bookmark=id.q68srxklvpt4.
+  rpc CreateJobExecution(CreateJobExecutionRequest) returns (JobExecution) {
+    option (google.api.http) = {
+      post: "/v1/executions"
+      body: "execution"
+    };
+  }
+  rpc GetJobExecution(GetJobExecutionRequest) returns (JobExecution) {
+    // Client example:
+    //   curl http://DOMAIN_NAME/v1/executions/1
+    option (google.api.http) = {
+      get: "/v1/executions/{execution}"
+    };
+  }
+  rpc ListJobExecutions(ListJobExecutionsRequest) returns (JobExecutions) {
+    // Client example:
+    //   curl
+    //   http://DOMAIN_NAME/v1/executions?job_name=my-prow-job&status=SUCCESS
+    option (google.api.http) = {
+      get: "/v1/executions"
+    };
+  }
+}
+
+message CreateJobExecutionRequest {
+  string job_name = 1;
+  JobExecutionType job_execution_type = 2;
+  GitReferenceDynamic git_refs = 3;
+  PodSpecOptions pod_spec_options = 4;
+}
+
+message PodSpecOptions {
+  map<string, string> envs = 1;
+  map<string, string> labels = 2;
+  map<string, string> annotations = 3;
+}
+
+/* Look up a single Prow Job execution. */
+message GetJobExecutionRequest {
+  string id = 1;
+}
+
+/* Look up all Prow Job executions that match all fields given here. */
+message ListJobExecutionsRequest {
+  string job_name = 1;            // Mapped to URL query parameter `job_name`.
+  JobExecutionStatus status = 2;  // Mapped to URL query parameter `status`.
+}
+
+message JobExecutions {
+  repeated JobExecution job_execution = 1;
+}
+
+message JobExecution {
+  string id = 1;
+  CreateJobExecutionRequest create_job_execution_request = 2;
```

Comment:
The reasoning for this was because the CreateJobExecutionRequest was only used by the web API. However, now that Sub is also converting its ProwJobEvent to a CreateJobExecutionRequest, it probably makes sense to keep this object as-is. Can you think of other reasons why we would want to flatten this out?

For some deeper pros/cons of nesting vs flattened structure, see https://stackoverflow.com/questions/49670346/is-it-best-to-have-many-field-in-protobuf-message-or-nested-messages. I personally agree with the accepted answer there that keeping the object model intact (keeping this nested) is more important than the performance penalties that come with it.